gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -86,9 +86,13 @@ public CloudStorageFileSystemProvider() {
     this(storageOptions);
   }
 
-  private CloudStorageFileSystemProvider(@Nullable StorageOptions gcsStorageOptions) {
+  CloudStorageFileSystemProvider(@Nullable StorageOptions gcsStorageOptions) {
     if (gcsStorageOptions == null) {
-      this.storage = StorageOptions.defaultInstance().service();
+      if (storageOptions == null) {
+        this.storage = StorageOptions.defaultInstance().service();
+      } else {
+        this.storage = storageOptions.service();
```

Comment:
The potentially confusing scenario is that forBucket accepts options (it must, to allow for user-provided credentials) and that CloudStorageFileSystemProvider allows for default options to be set (we're using this to set up the fake for testing). We need default options if we want to be able to test the automatic ways in which the filesystem can be created. We need explicit options if we want to allow the user to be able to set up auth. Having both default options and explicit options, one has to have precedence over the other.

The code you highlighted is really more of a collateral victim of this underlying dual function rather than the cause. In its defense, it's package-internal.

I would argue that the best thing to do at this point is to leave the code as it is (using the default options unless explicit options are provided), and revisit once @jart changes the way we introduce the fake (though I can add comments to clarify what's going on). I'm open to other ideas though if you have a proposal.